Add gce instance authority

matt0bi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines <https://github.com/spotify/gordon-janitor-gcp/blob/master/CONTRIBUTING.rst>
2. If the PR is unfinished, please prefix the subject line with [WIP], [DRAFT], or [RFC].
-->

**What this PR does / why we need it**:

This PR adds the code required to pull instance data for each GCE project in an organization and send it data to the `Reconciler` for further processing. 

**Which issue(s) this PR fixes**
<!-- optional; in `fixes #<issue number>, fixes #<issue_number>, ...` format, will close the issue(s) when PR gets merged: -->
N/A

**Special notes for your reviewer**:

First off, I apologize for the size of this PR. It's been a long time in the making and underwent multiple revisions while other code was being merged into the `develop` branch. Because of the same reason, this PR includes some things that will be improved in a follow-up PR since I think it's better to merge this now instead of keeping around such a long-lived branch even longer.

Here is a short, but not exhaustive list of items that will be followed up:
- Some fixtures look like they could be collapsed into one and shared between tests. For example, different AIOGoogleHTTPClient subclass tests set up a very similar mock auth client fixture.
-  There's some duplicated pagination logic between the DNSClient and the two clients in the `gce_clients.py` module that could be de-duped.
- It would be good to leave some comments about where to emit monitoring data.

There are also a few things that I'm unsure about that I'll comment on.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, start the release note with the string "action required: ".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

@spotify/alf @econchick

matt0bi committed:
Extract common test helpers

matt0bi committed:
Add msg_generators module

matt0bi committed:
Add gce_clients module

matt0bi committed:
Add gce_instance_authority module

matt0bi commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+        items = []
+        next_page_token = None
+
+        while True:
```
I've added some basic retry logic here, but it feels like it would simpler to have that logic in the client itself or make use of a 3rd party module. If we decide to use our own logic, we should change this to an exponential backoff and introduce timeouts to the requests since we've been bitten by this before.

matt0bi commented on gordon_janitor_gcp/msg_generators.py:
```diff
@@ -0,0 +1,93 @@
+"""
+
+
+class MsgGeneratorBase:
```
The reason why I decided to cordon off message creation logic in its own module is that a lot more of it will come in order to handle our company specific record setup (multiple types of records, some business logic). 

The `ARecordMsgGenerator` is more of an example implementation than anything else since we won't use it.

matt0bi commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+from gordon_janitor_gcp import gce_clients
+
+
+class TestPagingGoogleClientBase:
```
This test module uses test classes in order to get some feedback about how others feel about using test classes. It's a nice way to group things together, but it makes using `pytest.mark.parametrize` somewhat more difficult to use and adds more complexity over using test functions like we've always done.

Once I get some feedback, I'll either rework this into simple test functions or see about making other tests into classes.

econchick replied:
On your test methods, can you have a docstring detailing what is being tested? I think you could use these docstrings as a way to shorten your test method names.

matt0bi commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+    def _extract_instance_data(self, instance):
+        return {
+            'hostname': instance['name'],
+            'internal_ip': instance['networkInterfaces'][0]['networkIP'],
```
I used square brackets in this instances because I wanted to catch, log, and ignore errors where incomplete instance data was returned, but if someone knows a better approach then I'm game.

econchick commented on gordon_janitor_gcp/msg_generators.py:
```diff
@@ -0,0 +1,93 @@
+    def __init__(self, zone):
+        self.zone = zone
+
+    async def generate_msg(self, instance):
```
I know the base class is async, but does it need to be?

matt0bi replied:
I think it makes things simpler by having the interface stay the same (always async). Do you think this is a strong enough point or is it premature optimization?

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Client modules to to retrieve project and instance data from GCE.
```
s/modules/module, and there's an extra `to` in there :)

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+        instances = await client.list_instances('project-id')
+        print(instances)
+
+    loop.run_until_complete(main())
```
Don't forget to create a `loop` beforehand.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+                    msg = base_msg + f'RETRYING. {retries} retries left.'
+                    logging.error(msg)
+                    await asyncio.sleep(retry_wait)
+                    continue
```
Maybe I'm not properly caffeinated - but does this actually work?  When continuing, you don't have a `response` object to use when calling `self.parse_response_into_items`.

matt0bi replied:
When continuing, the execution jumps back to the beginning of the loop, so `self.parse_response_into_items` won't get called. If a response object is never created because all the attempts fail, the function will return in the `else` clause.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+from gordon_janitor_gcp import http_client
+
+
+class PagingGoogleClientBase(http_client.AIOGoogleHTTPClient):
```
Any reason for this not to live in the `http_client` module?

matt0bi replied:
~Good idea. How do you feel about combing this with its base class? Not every client will need to paginate everything, but in that case the inheriting class doesn't need to use the functionality.~

Edit 02/20: I moved this into `http_client` and renamed it to `PagingGoogleClientMixin` because I think it helps separate responsibilities nicely _and_ makes things more composable.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+        return {
+            'hostname': instance['name'],
+            'internal_ip': instance['networkInterfaces'][0]['networkIP'],
+            'external_ip': instance['networkInterfaces'][0]['accessConfigs'][0]['natIP'],  # noqa: E501
```
To avoid flake8 skipping this, why don't you just do something like:

```python
def _extract_instance_data(self, instance):
    iface = instance['networkInterfaces'][0]
    return {
        'hostname': instance['name'],
        'internal_ip': iface['networkIP'],
        'external_ip': iface['accessConfigs'][0]['natIP'],
    }
```

matt0bi replied:
_Brilliant_

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+        Return:
+            bool: if instance has any blacklisted tag.
+        """
+        instance_tags = instance['tags'].get('items', [])
```
Will there always be a `'tag'` key? Otherwise you're not handling a potential `KeyError`.

matt0bi replied:
I copy-pasta'd it from zonextgen, but it's a good point - I changed it to `instance.get('tags', {}).get('items', [])`

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+        Return:
+            bool: if instance has blacklisted metadata.
+        """
+        instance_metadata = instance['metadata'].get('items', [])
```
Same here - will there always be a `'metadata'` key?

matt0bi replied:
Switched it to `instance.get('metadata', {}).get('items', [])`

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+        for metadata in instance_metadata:
+            for bl_meta in self.blacklisted_metadata:
+                if bl_meta.get(metadata['key']) == metadata['value']:
+                    msg = (f'Instance "{instance["name"]}" filtered out for '
```
I think you might need to escape a set of double-quotes here.

matt0bi replied:
Tried this out in ipython, it actually works thanks to some f-string magic:

```python
In [1]: interpol = {'name': 'Bob'}

In [2]: print(f'The cat\'s name is "{interpol["name"]}"')
The cat's name is "Bob"
```

econchick replied:
yeah I think I was just being an idiot here 🙃

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+            def parse_response_into_items(self, response, items):
+                items.append(response['data'])
+
+        client = get_gce_client(TestClient)
```
nitpick, you can just return from here.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        client = get_gce_client(TestClient)
+        return client
+
+    def test_parse_response_into_items_raises_when_not_implemented(
```
I think you can get rid of the word `when` to make this a little shorter.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        with aioresponses() as m:
+            m.get(
+                'http://example.com',
+                exception=exceptions.GCPHTTPError('504 Gateway Timeout'))
```
To make this more readable, I would separate this out a bit, like so:

```py
exc = exceptions.GCPHTTPError('504 Gateway Timeout')
with aioresponses() as m:
    m.get('http://example.com', exception=exc)
```

But, I think rather than having aioresponses raise an exception, it should return the status code (more likely to happen than the server itself giving you an exception), e.g. `m.get('http://example.com', status=504)`

matt0bi replied:
I was able to sidestep this after I pulled out the paging functionality into `http_client` - I still don't know why I was getting the problem I described to you on Friday, but I'll dig into it some other time.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        retry_wait = 5
+        with aioresponses() as m:
+            m.get(
+                'http://example.com',
```
Any reason not to use `conftest.API_BASE_URL`?

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+            # second call fails despite retries
+            m.get(
+                'http://example.com/?pageToken=token',
+                exception=exceptions.GCPHTTPError('504 Gateway Timeout'))
```
Same thing re: `status=504` here.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+            # One successful call
+            m.get(
+                'http://example.com',
+                payload={'data': 'data', 'nextPageToken': 'token'})
```
Same thing here re: making it more readable:

```py
payload = {'data': 'data', 'nextPageToken': 'token'}
with aioresponses as m:
    m.get('http://example.com', payload=payload)
```

econchick replied:
Also, the `list_all_items` methods call `get_json`, but what is actually returned to the `get_json` method is text, not json. Since we're mocking the response (and not what `get_json` returns), it should be something like: 

```python
payload = {'data': 'data', 'nextPageToken': 'token'}
body = json.dumps(payload)
with aioresponses as m:
    m.get('http://example.com', body=body)
```

matt0bi replied:
:+1: Made it more readable. Regarding the payload part - aioresponses [already does that automagically](https://github.com/pnuckowski/aioresponses/blob/4d94c146172a085664b72363c53491e39bb871dc/aioresponses/core.py#L32)

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+class TestCRMClient:
+    @pytest.fixture
+    def crm_base_url(self):
+        return 'https://cloudresourcemanager.googleapis.com/v1/projects?'
```
I would not use the real URL for fear of accidentally improperly mocking it out and making a real call.

matt0bi replied:
Great catch! I've changed it so it's a fake url and the `CRMClient.BASE_URL` attribute gets patched with it.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+                    'labels': {},
+                    'parent': {}
+                }],
+            'nextPageToken': None
```
I didn't think these were present if there is no token, but maybe I'm wrong.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+    async def test_list_all_active_projects_multiple_pages(
+            self, mocker, crm_one_page_rsp, crm_base_url, get_gce_client):
+        crm_client = get_gce_client(gce_clients.CRMClient)
+        page2 = copy.deepcopy(crm_one_page_rsp)
```
Is `deepcopy` needed? I thought it was only when objects contain other objects. I think just `crm_one_page_rsp.copy()` would be sufficient.

matt0bi replied:
I thought the same, but the nested lists and dicts inside of `crm_one_page_rsp` are objects, so using shallow copies meant that I was modifying the same dicts.

econchick replied:
eww okay

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        with aioresponses() as m:
+            m.get(
+                crm_base_url + 'pageSize=500',
+                payload=crm_one_page_rsp)
```
Same thing here (and below) re: `payload` vs `body`.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+class TestGCPClient:
+    @pytest.fixture
+    def compute_base_url(self):
+        return ('https://www.googleapis.com/compute/v1/projects/'
```
Same thing here re: using real URLs for tests.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+                }
+            },
+            'nextPageToken':
+            None
```
If `'nextPageToken'` key is indeed there even with no pages, then I think you have an extraneous new line here.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+                page_size=10,
+                instance_filter='zone eq us-west9-z')
+
+            expected_results = [{
```
I don't think this needs to be within the `with aioresponses() as m` context, and can be outside.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        gcp_client = get_gce_client(gce_clients.GCPClient)
+        caplog.set_level(logging.ERROR)
+        blacklisted_instance = \
+            copy.deepcopy(compute_rsp['items']['us-west1-z']['instances'][0])
```
Same thing here with deepcopy vs copy.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+            results = await gcp_client.list_instances(
+                'test-project', page_size=10)
+
+            expected_results = [{
```
Same thing here being outside of the `with` context.

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        assert len(caplog.records) == 1
+
+    @pytest.mark.asyncio
+    async def test_list_instances_filter_by_metadata(
```
Can this test + `test_list_instances_filter_by_tag` + `test_list_instances_from_one_page` be parametrized at all? Seems like a lot of repeated code.

matt0bi replied:
:+1: Packed all of them down into one test!

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+    @pytest.mark.asyncio
+    async def test_list_instances_filter_by_metadata(
+            self, compute_rsp, compute_base_url, get_gce_client, caplog):
+        gcp_client = get_gce_client(gce_clients.GCPClient, )
```
Is the `, ` needed?

econchick commented on tests/unit/test_gce_clients.py:
```diff
@@ -0,0 +1,344 @@
+        assert len(caplog.records) == 1
+
+    @pytest.mark.asyncio
+    async def test_list_instances_malformed_instance_json(
```
This might also be able to be included in the parametrization suggested above.

matt0bi replied:
I got this working, but the resulting test was harder to understand because this case is different from the previous three and required a lot of setup specific to just this one case.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+                if retries > 0:
+                    retries -= 1
+                    msg = base_msg + f'RETRYING. {retries} retries left.'
+                    logging.error(msg)
```
I think this should be a warning, not an error.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+                    continue
+                else:
+                    msg = base_msg + f'NO RETRIES left, return partial reply'
+                    logging.error(msg)
```
Same thing - a warning, particularly if you are returning items / not raising.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+                    try:
+                        items.append(self._extract_instance_data(instance))
+                    except KeyError as e:
+                        logging.error(
```
Since you're continuing on and not crashing, this should be a warning.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+            if tag in self.blacklisted_tags:
+                msg = (f'Instance "{instance["name"]}" filtered out for '
+                       f'blacklisted tag: "{tag}"')
+                logging.error(msg)
```
Any reason this can't be `info`? I don't even think it constitutes as a `warning`.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,297 @@
+                if bl_meta.get(metadata['key']) == metadata['value']:
+                    msg = (f'Instance "{instance["name"]}" filtered out for '
+                           f'blacklisted metadata: "{bl_meta}"')
+                    logging.error(msg)
```
Same here - probably more like an `info` level message.

matt0bi replied:
Thanks for pointing all the `logging` related issues, this makes a lot more sense.

econchick reviewed (changes requested):
Eep, sorry for so many comments. This is just first pass too - I want to look a bit more deeply at the logic itself.

matt0bi commented:
Hey @econchick, I've addressed your feedback and rebased off of develop - could you please take another look when you have a chance?

Once this PR is in good shape, I'll close it, push the commits to `feature/add_gce_instance_authority`, open a new PR, and merge that one in order follow your branch naming suggestion.

descent3me commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,234 @@
+
+    async def main():
+        session = aiohttp.ClientSession()
+        loop = asyncio.get_event_loop()
```
are there two loops now?

descent3me commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+    >>> rrset_channel = asyncio.Queue()
+    >>> authority = GCEInstanceAuthority(config, rrset_channel)
+    >>> await authority.start()
+    >>> msg = await.channel.get()
```
is this the correct syntax? i'm not entirely familiar with asyncio api, but it looks a bit odd

descent3me commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+        Returns:
+            str: filter suitable for the 'filter' endpoint parameter.
+        """
+        return ''
```
is there a reason to have this as a method, as opposed to as an instance variable?  why does it need to be async?

econchick replied:
Yeah I was going to ask about this - why is it even needed?

econchick replied:
And if it is needed, maybe it can be a private method (if it has to be a method)

matt0bi replied:
Good q's. I set this up this way so that gordon can read some pods from the config (because pods can be read and understood by humans), contact podservice to translate the pods into google zones (eg. 'pod1' -> 'us-central1-z'), and use that zone to build a query string that's understood by the API endpoint (eg. `zone eq .*/us-central1-z'). It returns an empty string because the pod-zone-filter logic is specific to Spotify and, afaik, shouldn't be in this repo.

If I'm thinking this through right, removing this would mean either having the same logic in a different place or having us configure each instance of the janitor with concrete zone names. Does this sound right?

> And if it is needed, maybe it can be a private method (if it has to be a method)

Good idea, I'll change it to be private.

econchick replied:
I guess I'm confused at how non-Spotify users are suppose to make use of this filter...

descent3me commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+        :py:obj:`self.rrset_channel`.
+        """
+        async with aiohttp.ClientSession() as session:
+            self.session = session
```
I'm interested in why this is set with an async context manager, which I also saw in the all the examples in the docs, but I couldn't find the reasoning for it. I understand session requests need to be async, but not necessarily why establishing the session needs to be

econchick replied:
I had a similar/related question too - but it's also a question of the bigger picture. I know that using a context manager with `aiohttp.ClientSession` automatically closes the session once you're done with the context/outside the scope. But one could also directly close the client session upon cleanup of the client (i.e. emit a `None` message after all messages sent, session._close).

I initially was curious why creating a new client session for this, since an auth client will create one, and the http client will automatically reuse the auth client one. I then realized that because of all the iteration going on here (`async for project_instances in self.get_instances_for_projects`, then `for project in projects`, then iterating over pages, then all instances of the complete response - loooots of iteration), and sort of forcing you to create a new instance of the auth client, and GCE client, and the CRM client. So then the questions has morphed into: can't we create one instance of the three clients, and reuse them all? I don't think these new instances of clients are different every time (unless I'm completely missing on something). So no need for creating an explicit session, no need for a new object every iteration. Could the three just be created upon instantiation of the Authority class? Similar to how the reconciler currently does it, and the proposed pubsub client does.

matt0bi replied:
I think the reason why this is using an `async with` statement is because closing the session is an async operation, so an ordinary context manager would not be able to call the session's [async close](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L696) method.

Why use a context manager at all here? I think it's a choice between that and a `try...except...finally` block because it'd be nice if the authority closed the session when it finishes or when an uncaught exception occurs. I couldn't think of anything the try-except block could catch, except maybe a generic `except Exception`, so I went with `with`. 

> can't we create one instance of the three clients, and reuse them all?

Can you point out where a new object is created on very iteration? In my understanding, there are only 4 client-like objects: 2 auth clients, 1 CRMClient, and 1 GCEClient. Each client is reused after instantiation, and each of them uses the same `aiothttp.ClientSession`.
Also, can one auth client be shared between clients that use different scopes and service accounts?

This made me notice that we never await `session.close()` calls in the reconciler. Not entirely sure what this means, but could be something like leaving sockets open, so we should patch that up later.

econchick replied:
Ah you're right re: not multiple clients - that's what I get for reviewing while jet lagged. It was just confusing with the generator. 

Good call on the `await` missing. I don't think I'll have time to make a quick PR to fix so if you want to, by all means :)

I do think the argument for init'ing all the clients when the authority plugin is init'ed is probably still valid, though, right?

Gah I should get some sleep.

descent3me replied:
Thanks for the explanation to my question! I had also forgotten that the way we plan to have multiple janitors in one process. I was thinking the program would end immediately after start() was called, and since we've toyed with the idea of running this on a 5 min cron, I figured if any resources were claimed, they would be freed by then, so I didn't really see the point of calling close (I may have some other misconceptions there, too, but it seems like close doesn't do *that* much https://aiohttp.readthedocs.io/en/stable/client_advanced.html#graceful-shutdown). It was also my understanding that close() was not an async method, since you can call it synchronously and the docs don't mention one way or the other. However, now that I've looked at the source code (https://github.com/aio-libs/aiohttp/blob/master/aiohttp/client.py#L696), I see that it is defined as async; it just doesn't raise any exceptions if you call it synchronously.

matt0bi replied:
> I do think the argument for init'ing all the clients when the authority plugin is init'ed is probably still valid, though, right?

@econchick In this instance, initing the clients in specific methods makes it easy to override these methods and put some internal-only logic into them - if I moved the client inits to the plugin's init, then adding internal logic would require more changes (eg. would have to override __init__(), the method in question, and possibly start()). 

This assumes that the internal-facing plugin package will depend on this package so that it can import and override some things. Is there some other way to handle the public/internal split that I'm missing? We can talk about it on slack if that makes it easier.

matt0bi replied:
@descent3me I think that's a good point. If I remember correctly, the aim is to kick off the janitor and let it run for a few minutes to do its job, in which case you're right that we could leave this stuff hanging and trust the OS to clean up any sockets. I'll think a bit more what this means if we have >1 ClientSessions and see which way to go.

I dug in deeper into the `close` stuff, and [it never does anything asynchronous](https://github.com/aio-libs/aiohttp/blob/master/aiohttp/connector.py#L314) later as far as I can tell, so there's no need for the async context manager, unless the library author is planning something for the future. That said, if an async method gets called but not awaited, it never gets scheduled for execution and the interpreter issues a warning, so all those calls to `session.close()` are noops :):

```pycon
In [6]: import asyncio

In [7]: async def work():
   ...:     print('working')
   ...:     await asyncio.sleep(1)
   ...:     
   ...:     

In [8]: async def main():
   ...:     work()
   ...:     await work()
   ...:     

In [9]: loop = asyncio.get_event_loop()

In [10]: loop.run_until_complete(main())
/home/matto/.virtualenvs/sandbox/bin/ipython:2: RuntimeWarning: coroutine 'work' was never awaited
  
working
```

descent3me reviewed (commented):
I didn't get all the way through either, but I saw a couple examples that may need updating, and a question about aiohttp sessions

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,234 @@
+        return await self.list_all_items(url, params, retries, retry_wait)
+
+
+class GCPClient(http_client.AIOGoogleHTTPClient,
```
Should be named `GCEClient` since it's interacting with compute engine, rather than *all* of GCP.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,234 @@
+        """Fetch all instances in a GCE project.
+
+        Endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
```
Can you make this a link with rST? e.g. 

```rst
You can refer to the `endpoint documentation here <https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList>`_.
```

(looks like github pygments/code highlighting might not work for rST 😬 )
A line break _should_ just be ignored.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,234 @@
+        """
+        projects = response.get('projects', [])
+        for project in projects:
+            if project.get('lifecycleState', '').lower() == 'active':
```
This feels weird to me that the filtering of active projects is in `parse_response_into_items` method rather than `list_all_active_projects` method. I think it might be more indicative of some code smell, like you're forcing the pagination client to work for this CRMClient.

matt0bi replied:
I don't see anything weird about it, it looks like a regular ol' mixin class. But since it can cause confusion, I rewrote the pagination mixin and these clients to make 'em less confusing.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,234 @@
+            'external_ip': iface_data['accessConfigs'][0]['natIP'],
+        }
+
+    def _blacklisted_by_tag(self, instance):
```
With this and with the metadata one, I'm wondering if you can push some filtering to the API call with their filter query parameter 🤔

matt0bi replied:
I tried to use that feature, but I couldn't get it it filter on nested items like tags and metadata. The example the docs give even includes a nested structure, but that didn't help.

econchick commented on gordon_janitor_gcp/http_client.py:
```diff
@@ -169,3 +170,59 @@ def __init__(self, auth_client=None, session=None):
+                response = await self.get_json(url, params=params)
+            except exceptions.GCPHTTPError as e:
+                base_msg = f'Fetching items from "{url}" failed: {e}, '
+                if retries > 0:
```
I think retry logic should actually live in the `AIOGoogleHTTPClient`, as well as allow retries to be configurable when instantiating the http client. It could live along side the refresh attempts (and maybe separated out into each of their own methods? if that could work?).

matt0bi replied:
I've decided to remove this altogether and leave it for another since it's not part of the ticket.

econchick commented on tests/unit/test_http_client.py:
```diff
@@ -223,3 +223,87 @@ def simple_json_callback(resp):
+            # second call fails despite retries
+            raise exceptions.GCPHTTPError('504 Gateway Timeout')
+
+    mocker.patch.object(simple_paging_client, 'get_json', get_json)
```
Is this the way around trying to get `aioresponses` to work the way you want to?

matt0bi replied:
Not really, it's because the mixin is easier to test when it's isolated, so it doesn't need to make http requests.

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+        """Get filter string used by instances.aggregatedList endpoint.
+
+        For filter string details, see api endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
```
Can you also make this a clickable rST link too?

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+            list: project IDs.
+        """
+        projects = await self.get_all_projects()
+        project_blacklist = set(self.config['project_blacklist'])
```
Unhandled potential `KeyError` here.

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+        active_projects = await crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def get_processed_projects(self):
```
I find this to be a weird method name. `processed` sounds like you're doing something to the projects, when in actuality, you're filtering projects (active + blacklist). would `get_projects` be sufficient?

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+        Returns:
+            obj: obj that creates messages for the channel.
+        """
+        return msg_generators.ARecordMsgGenerator(self.config['zone'])
```
A couple of things: this is an unhandled `KeyError` if `'zone'` is not in config. 

And, following the `msg_generator` logic, the zone in config is the DNS zone. But I think I am not understanding how the instances are actually filtered for the given zone. 

That also got me thinking (and I could be completely misunderstanding your code - I'm so sorry if I am): if we're configuring for one DNS zone, then we'll need multiple janitors to run, one for each DNS zone. Perhaps there can be a list of zones provided to the authority, so then all instances of all projects are polled, then divvied up into the batched messages. 

I'm hoping/thinking that the list of zones could almost be a global gcp plugin janitor config item, but it might be Authority client specific, or Reconciler specific (i.e. Cloud DNS refers to zones as `managedZones` - where should that go in config? does this name even matter if the DNS client knows what we're talking about when we just say `zones`?

econchick replied:
Also, I assume `PTR` and `CNAME` records are generic need to address. Half-baked/jet lagged idea: perhaps this can be configurable, e.g. `record_types_to_generate` or `record_types_to_reconcile` or just `record_types` on the global plugin level:

```toml
[gcp]
record_types = ["A", "CNAME", "PTR"]
```

What do you think?

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+    """
+    def __init__(self, config, rrset_channel=None, **kwargs):
+        self.config = config
+        self.keyfile = self.config['keyfile']
```
Another unhandled `KeyError`. It should be the responsibility of the plugins to handle config, including erroring out if a required config item is needed, and should probably be done as early as possible - i.e. when plugins are loaded & initialized with the plugin-specific config. For example, the reconciler logic [errors out if missing a `keyfile` key](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/gdns_reconciler.py#L105-L111) as well as [a project key](https://github.com/spotify/gordon-janitor-gcp/blob/develop/gordon_janitor_gcp/gdns_reconciler.py#L119-L124).

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+        """
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile,
+            scopes=['cloud-platform.read-only'],
```
For this and the other part where scopes are passed in - is there a reason to hard code and not have it optionally configurable?

Also is there a non-`read-only` need? could `cloud-platform.read-only` be the default for all (exception of publishing to pubsub).

econchick commented on gordon_janitor_gcp/gce_instance_authority.py:
```diff
@@ -0,0 +1,145 @@
+            yield await gcp_client.list_instances(
+                project, instance_filter=instance_filter)
+
+    async def get_msg_generator(self):
```
Any reason for this to be a public method? This is actually a question for the overall PR.

econchick commented on gordon_janitor_gcp/gce_clients.py:
```diff
@@ -0,0 +1,234 @@
+from gordon_janitor_gcp import http_client
+
+
+class CRMClient(http_client.AIOGoogleHTTPClient,
```
This is more of a nitpick, but I'm wondering if we should standardize on naming. There's `GDNSClient`\*, `GCPClient` (though should be renamed to `GCEClient`), and then `CRMClient` - maybe `GCRMClient` and something like `GPaginatorMixin`, or something else?

\* FTR I have a local WIP patch that renames `GoogleDNSReconciler` to `GDNSReconciler`, `GoogleAuthClient` to `GAuthClient`, `AIOGoogleDNSClient` to `GDNSClient`, `AIOGoogleHTTPClient` to `AIOConnection`, and `GooglePubsubPublisher` (still the open PR) to `GPubsubPublisher` (which is both the publisher interface and API client of pub/sub).

Thoughts?

econchick reviewed (changes requested):
Eeep - lots of comments, so sorry 😬 

In reflecting on some of my comments, I feel like I might be pushing my approach to plugin development onto you (config verification upon plugin instantiation, auth/API client instantiation upon plugin instantiation, separate `done` method to "clean up", http client/pagination, etc). I hope you know that if you don't like/agree with that, I'm totally up for talking it through to agree it, since we should align and have consistency. And also documenting it for other/future plugin development.

matt0bi commented:
I'm closing this PR in order to break it up into smaller ones. I've created #9 that contains the gcp-related code.
